src/main/scala/com/databricks/labs/overwatch/env/Database.scala

```diff
@@ -239,14 +241,65 @@ class Database(config: Config) extends SparkSessionWrapper {
         spark.streams.removeListener(streamManager)
 
       } else { // DF Standard Writer append/overwrite
-        target.writer(finalDF).asInstanceOf[DataFrameWriter[Row]].save(target.tableLocation)
+        try {
+          target.writer(finalDF).asInstanceOf[DataFrameWriter[Row]].save(target.tableLocation)
+        } catch {
+          case e: Throwable =>
+            logger.log(Level.ERROR, s"""Exception while writing to ${target.tableFullName}""""+Thread.currentThread().getName)
+            throw e
+        }
       }
       logger.log(Level.INFO, s"Completed write to ${target.tableFullName}")
     }
     registerTarget(target)
     true
   }
 
+  def writeWithRetry(df: DataFrame,
+                     target: PipelineTable,
+                     pipelineSnapTime: Column,
+                     maxMergeScanDates: Array[String] = Array(),
+                     daysToProcess: Option[Int] = None): Boolean = {
+
+    var inputDf = df
+    if (daysToProcess.nonEmpty) {
+      if (daysToProcess.get < 5 && !target.autoOptimize) {
+        logger.log(Level.INFO, "Persisting data :" + target.tableFullName)
+        inputDf = df.persist()
+        inputDf.count()
+      }
+    }
```

Comment:
why are you creating another var here? and why do you have a nested if statement? what if daysToProcess.isEmpty -- what then? If that's part of the sub-if then just add it as another and statement.

should be a val -- potential solution is below

```scala
  val needsCache = daysToProcess.getOrElse(1000) < 5 && !target.autoOptimize
  val inputDf = if (needsCache) {
      logger.log(Level.INFO, "Persisting data :" + target.tableFullName)
      df.persist()
    } else df
  
  if (needsCache) inputDf.count()
```

Reply:
done

Reply:
good job!